### PR TITLE
Packet Trace CLI

### PR DESCRIPTION
**Pull request by s3wong**
Signed-off-by: Stephen Wong s3wong@midokura.com

### CODE REVIEW

**Comment by admytren** [comment]
HI @s3wong, could you post a sample of list/get/delete operations?

**Comment by s3wong** [comment]
@admytren 

list:

midonet> trace list
trace trace0 
trace trace1 
trace trace2 
trace trace3 
trace trace4 
trace trace5 
trace trace6 
trace trace7 
trace trace8 

show:

midonet> trace trace0 show
trace trace0 messages [u'2013-08-21 03:07:15.676 afeb432d-e309-4b87-9e14-43bea4c28577 Entering device', u'2013-08-21 03:07:15.711 0ed7c323-44fc-46e6-b344-ddccf52178cb Forwarded to port', 
u'2013-08-21 03:07:15.738 5a1a66de-12e3-4a33-a1a0-92a4534b1127 Entering device', 
u'2013-08-21 03:07:15.792 (none) Consumed']

delete:

midonet> trace trace4 delete
midonet> trace list
trace trace0 
trace trace1 
trace trace2 
trace trace3 
trace trace5 
trace trace6 
trace trace7 
trace trace8

**Comment by admytren** [comment]
@s3wong, Would it be possible to replace "traceX" in the list command with any useful information? For example, "trace trace0 to W.X.Y.Z" or "trace trace0 [timestamp]". Because in current format it's very unfriendly to the user. I don't know about Cassandra, but for Zookeeper ordering of traces always changes on CLI reload. And the customer may have to repeatedly show all traces until they find the one that's of interest to them.

**Comment by s3wong** [comment]
@admytren 

The cli generated object traceX is identified by 'id' (UUID) internally during the lifetime of midonet-cli session, and since UUID is part of the key to fetch trace messages/index from Cassandra, within the lifetime of the midonet-cli session, it stays the same

That is not to say that your point on display being user-unfriendly isn't valid. What you suggested, however, does require (mostly) backend changes - on both API server as well as MM. Ideally, MM would write the trace message with multiple columns denoting different data (timestamp, IP or MAC address, message...etc); then API server should fetch each key from the index table ('trace list' is basically dumping the index table), then additional perform lookup on the message table to get the first entry and extract the information.

Given the scope, though, I would suggest making this an enhancement

**Comment by admytren** [comment]
Would it be possible to add code to CLI (not API) to add something to make it possible to differentiate traces in the list? For example, when listing them, grab the first message and include it in the output. E.g.:

trace trace0 2013-08-21 03:07:15.676
trace trace1 2013-08-20 11:22:01.341

**Comment by admytren** [comment]
The code looks good. I would really recommend making this feature usable. If there is absolutely no time, it would be good to file this as a usability bug and not an enhancement...

**Comment by s3wong** [comment]
@admytren 

Possible of course :-)

In current form, it is difficult to do - the messages in message table are nothing more than a String. So if the API server just grabs the first message, it would actually look something like this:

trace trace0 '2013-08-21 03:07:15.676 afeb432d-e309-4b87-9e14-43bea4c28577 Entering device'

Chopping off just the "UUID" and "actual message" part becomes a String editing work to be done by API server - which does not seem appropriate as it should be done either as a format/schema breakdown directly from the writer, or perhaps editing on the cli side (whose job is to make display look good)

I think it is OK to file a usability bug to track better display for list (and even show)

**Comment by gontanon** [comment]
I'll second @admytren's comment: the code looks good, it's very raw in terms of usability. Maybe the demo is the time/place to get opinions about the priority of the usability enhancements?

**Comment by gontanon** [comment]
I forgot about this: could you update the man pages in doc/ too with the new types/fields? Thanks!

**Comment by jlmjlm** [comment]
Hey folks, it is not clear why this is being held up, so can you distinguish the work needed before you'll pull this into master from the stuff you'd like to have in the future, but not needed for this commit?

**Comment by s3wong** [comment]
Man page added to pulled request

File: doc/midonet-cli-trace.1.ronn [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+## DESCRIPTION
+
+Trace commands display MidoNet's packet tracing results. The user needs
+to first enable packet tracing via the trace-condition(1) command, then
+can display the trace messages via the trace command set
+
+To display the trace messages, user would have to pick from a set of 
+trace ids via the output of [trace list], then do a show command to display
+all the trace messages associated with that trace Id. Users can also delete
+the trace messages they no longer wanted to store via the delete command
```
Comment (admytren) [inline doc/midonet-cli-trace.1.ronn]: Recommend changing "wanted" to "want".

File: doc/midonet-cli-trace.1.ronn [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+## DESCRIPTION
+
+Trace commands display MidoNet's packet tracing results. The user needs
+to first enable packet tracing via the trace-condition(1) command, then
+can display the trace messages via the trace command set
+
+To display the trace messages, user would have to pick from a set of 
+trace ids via the output of [trace list], then do a show command to display
+all the trace messages associated with that trace Id. Users can also delete
```
Comment (admytren) [inline doc/midonet-cli-trace.1.ronn]: Recommend changing "trace Id" to either "traceId" (to match the parameter name) or "trace id"/"trace ID" (to have more uniform capitalization)

**Comment by admytren** [comment]
Looks good!

**Comment by s3wong** [comment]
Review comments above fixed

File: doc/midonet-cli-trace.1.ronn (hunk 2 above)
Comment (jlmjlm) [inline doc/midonet-cli-trace.1.ronn]: You should never use "id" in prose, as that's a legitimate word with a distinct, unrelated meaning.
I prefer "trace ID" to "traceId".

**Comment by gontanon** [comment]
IMHO this can be merged, the code is good, it works and it's a valid starting point. I'd say it's the PM's call if and when we should address the rawness in usability.

Unless those usability fixes mean incompatible data format changes for what we store in Cassandra / ZK.

**Pull request merged**